fix($resource): remove (broken) support for promises as `timeout`

gkalpak opened the pull request:
This fixes #13393 by:
1. Reverting 7170f9d.
2. Update the docs to note that promise as timeout is not supported.
3. Log a warning when someone tries to use it.

The commits could/should be squashed before merging. We probably need a BC notice (since the broken support for promises as `timeout` has sneaked into `v1.4.8`, right ?

gkalpak committed:
revert: fix($resource): allow XHR request to be cancelled via timeout promise

This reverts commit 7170f9d9ca765c578f8d3eb4699860a9330a0a11.

Fixes part of #13393.

petebacondarwin commented on src/ngResource/resource.js:
```diff
@@ -150,8 +150,10 @@ function shallowClearAndCopy(src, dst) {
- *     should abort the request when resolved.
+ *   - **`timeout`** – `{number}` – timeout in milliseconds.<br />
+ *     **Note:** In contrast to {@link ng.$http#usage $http.config}, {@link ng.$q promises} are
+ *     **not** supported in $resource, because the same value has to be re-used for multiple
```
Same value **would** be used?

gkalpak replied:
I copied it from `v1.5.x`. I should change it there too, right ?

petebacondarwin replied:
If it makes sense to you.

gkalpak replied:
I updated it and changed it on master as well (3694390).

petebacondarwin commented on src/ngResource/resource.js:
```diff
@@ -568,17 +570,13 @@ angular.module('ngResource', ['ng']).
+                           '  Promises are not supported in $resource, because the same value ' +
+                           'has to be re-used for multiple requests.');
+              } else if (key != 'params' && key != 'isArray' && key != 'interceptor') {
+                httpConfig[key] = copy(value);
```
I think I find the switch statement easier to follow. What is the benefit of changing to if?

gkalpak replied:
I find the `switch` statement better, but it was added as part of 7170f9d (which was reverted).
I aimed for the least amount of change  - it's more obvious if you check the 3rd commit alone (https://github.com/gkalpak/angular.js/commit/ed915fe5bb15a5b7a53a8260e8e14fa7f1c29c6f).

(But the `switch` looks arguably better, so I am fine putting it back.)

petebacondarwin replied:
Let's use switch! But perhaps add it as a `refact` commit after the revert?

gkalpak replied:
Done.

petebacondarwin commented:
Some minor comments but lgtm

gkalpak committed:
refactor($resource): change if-block to switch-block for readability

gkalpak committed:
docs($resource): add note about _promises as timeout_ not being supported

Fixes part of #13393.

gkalpak committed:
fix($resource): don't allow using promises as `timeout` and log a warning

Promises never worked correctly as values for `timeout` in `$resource`, because the same value has
to be re-used for multiple requests (and it is not possible to `angular.copy()` a promise).
Now (in addition to ignoring a non-numeric `timeout`), a warning is logged to the console using
`$log.debug()`.

Partly fixes #13393.

gkalpak commented:
@petebacondarwin, I addressed the comments. What about a BC ?

Here's the situation:

- Up to `v1.4.7` (included), using a promise as `timeout` in `$resource`, would silently fail (i.e. have no effect).
- In `v1.4.8`, using a promise as `timeout` will have the (buggy) behaviour described in https://github.com/angular/angular.js/pull/12657#issuecomment-152108887. (I.e. it will work as expected for the first time you resolve the promise and will cancel all subsequent requests after that - one has to re-create the resource class. This is not documented.)
- With this change, using a promise as `timeout` will log a warning and ignore the `timeout` value.

Do we need a BC notice ?

---
It just occurred me: You said we should [_recommend an upgrade to 1.5_][1] in the docs and warning message. 

[1]: https://angular-team.slack.com/archives/angular-1/p1448615582000246

gkalpak commented:
I updated the docs and the warning message to add a recommendation to upgrade to `v1.5` if one needs support for cancellable `$resource` actions.

petebacondarwin commented:
Yes, we should put a simple BC notice along the lines of your comment https://github.com/angular/angular.js/pull/13462#issuecomment-162915050

I think in general it is safer to have BCs where they were not needed rather than the other way around

petebacondarwin commented:
Landed as 474865242c89ba3e8143f0cd52f8c292979ea730 and previous three commits.
